internal/integration/knot/knot_protocol.go

```diff
@@ -27,36 +30,92 @@ type protocol struct {
 	devices   map[string]entities.Device
 }
 
-func newProtocol(conf config.IntegrationKNoTConfig) (Protocol, error) {
+// Create a new connection with knot cloud
+func newProtocol(conf config.IntegrationKNoTConfig, deviceChan chan entities.Device, msgChan chan network.InMsg) (Protocol, error) {
 	p := &protocol{}
 
 	p.userToken = conf.UserToken
-
+	p.network = new(networkWrapper)
 	p.network.amqp = network.NewAMQP(conf.URL)
-	p.network.amqp.Start()
+
+	err := p.network.amqp.Start()
+	if err != nil {
+		log.WithFields(log.Fields{"integration": "knot"}).Error("Error connection to knot-cloud: %v", err)
+		return p, err
+	} else {
+		log.WithFields(log.Fields{"integration": "knot"}).Error("New connection to knot-cloud")
+	}
 
 	p.network.publisher = network.NewMsgPublisher(p.network.amqp)
 	p.network.subscriber = network.NewMsgSubscriber(p.network.amqp)
 
+	if err = p.network.subscriber.SubscribeToKNoTMessages(msgChan); err != nil {
+		log.WithFields(log.Fields{"integration": "knot"}).Error("Error to subscribe")
+		return p, err
+	}
+
+	p.devices = make(map[string]entities.Device)
+
+	go handlerKnotAMQP(msgChan, deviceChan)
+	go DataControl(deviceChan, p)
+
 	return p, nil
 }
 
+// Update the knot device information on map
+func (p *protocol) UpdateDevice(device entities.Device) error {
+
+	if _, d := p.devices[device.ID]; !d {
```

Comment:
What does variable d represent? It should have an easy to understand name.

Reply:
ok